### ISSUE TITLE
Selectively enforcing a "contents:" constraint according to a type parameter?

### ISSUE THREAD

**Issue by tmillican** [kaitai-io/kaitai_struct#432]
I'm doing some reversing work on a 3rd-party format that uses a typical magic bytes scheme in the file header. However, there are several instances of official (direct from said 3rd party) files that have busted headers with incorrect magic bytes (and incorrect header hashes, as well). The rest of the data is okay apart from the header size, but I can deal with that.

The application that uses these files seem to ignore the header signature and header size altogether, assuming the size to be a pre-ordained constant. I could do the same, but I would prefer to have a boolean `is_mangled` parameter for the `file_header` type. Downstream code can decide for themselves whether they care about the magic byte mismatch. If `is_mangled == false`, the parser will throw an exception if the bytes don't match.

I thought I could solve this with an instance in the `file_header` type. Something like:

    file_header:
      params:
        - id: is_mangled
          type: boolean
      seq:
        - id: raw_magic
          size: 10
        # <etc>
      instances:
        good_magic:
          value: "MAGICVALUE"
        expected_magic:
          value: '(is_mangled ? raw_magic : good_magic)'
        validate_magic:
          value: raw_magic
          contents: expected_magic

But this gives a compiler error:

    /instances/validate_magic/contents: invalid key found in value instance, allowed: doc, doc-ref, enum, if, value

For context: I'm invoking `ksc` by way of `ksv`. `ksv --version` reports `kaitai-struct-compiler 0.8`.

It sounds like the compiler objects to having a `contents` key in an `instance`. But according to the [instance spec](http://doc.kaitai.io/ksy_reference.html#spec-instance), any key valid for an `attribute` should also be valid for an `instance`, apart from `id`.

Am I missing something? If this approach won't work, is there something else that would?

**Comment by GreyCat** [comment]
If I followed you right, the original intent is to get signature check working conditionally, based on some sort of parameter provided into type (i.e. `is_mangled`).

I can think of 2 ways to do it. First, you can just do the comparison manually and provide end-user with a boolean result:

```yaml
file_header:
  seq:
    - id: raw_magic
      size: 10
    # <etc>
  instances:
    is_magic_ok:
      value: 'raw_magic == [77, 65, 71, 73, 67, 86, 65, 76, 85, 69]' # "MAGICVALUE"
```

There is little point in providing `is_mangled` then, as one can just ignore results of check, or check `is_mangled or is_magic_ok`.

Second way is to have two magic fields with `if` condition:

```yaml
file_header:
  params:
    - id: is_mangled
      type: boolean
  seq:
    - id: checked_magic
      contents: MAGICVALUE
      if: not is_mangled
    - id: raw_magic
      size: 10
      if: is_mangled
    # <etc>
```

This one would throw exception, if `is_mangled` is false and thus check is requested.

**Comment by GreyCat** [comment]
> Regarding the error, though, is `contains` not supported on `instance`? If that's the case, then the documentation for `instance` is incorrect/out-of-date.

Spec is definitely out of date, but probably not in the way you'd suspect. Actually, right now we have 2 types of instances:

* "parse instances", i.e. instances that actually do parsing starting at some arbitrary position in the stream. These must have `pos`, allow to specify `io`, and all other tags associated with attribute spec, including `contents`. You can, for example, do

```yaml
instances:
  foo:
    pos: 123
    contents: [1, 2, 3]
```

* "value instances" are pure calculated values which have no representation in the stream. Thus, they only can have `value` and `enum`, and, by design, it's pointless to have any constraints on it.

**Comment by tmillican** [comment]
Ah, that explains it. I've used both types, but I didn't realize that they were technically different things on the back end.

I think I'll just remove the constraint on the magic bytes, then, and leave a "caveat emptor" in its place. For my purposes (extraction/modification utility) I can always "taste" the magic bytes first before running the Kaitai-generated parser and at least warn people if they're trying to parse a very large file that isn't verifiably of the right type. It's kind of an unlikely scenario to begin with.

Katai Struct is amazing, by the way. I've been able to do in a couple days what would have taken at least a couple of weeks when you figure in all the testing, and much of that was just spent learning the ins-and-outs of Kaitai. It's honestly kind of appalling how many variations there are on the lex/yacc theme, and yet how little attention has been given to reasoning about other kind of computer languages that don't yield to that approach. When I get done with this little project (or done enough), I'll definitely be contributing my efforts to the format library.
